Implement cost provider and savings estimator for Databricks Azure platform

cindyyuanjiang opened the pull request:
Fixes #364 

This PR adds the cost savings (cost provider and savings estimator) implementation for Databricks Azure user qualification tool.

The Databricks Azure instance pricing information was pulled from: https://azure.microsoft.com/en-us/pricing/details/databricks/. We currently assume that the region is 'West US'.

cindyyuanjiang commented:
Now we can run Databricks Azure user qualification tool end-to-end. We tested the tool with the following script:

```
# assume there is a local work-dir called $SCRATCH_DIR where
# all the resources are being saved and a remote work-dir on ABFS (Azure Blob File System)


export SCRATCH_DIR=
export LOCAL_OUTPUT_DIR=$SCRATCH_DIR/output_folder
export CPU_CLUSTER_NAME=
export CPU_CLUSTER_PROPS=$SCRATCH_DIR/cluster_properties/cpu_cluster_props.json
export GPU_CLUSTER_NAME=
export GPU_CLUSTER_PROPS=$SCRATCH_DIR/cluster_properties/gpu_cluster_props.json

export REMOTE_SCRATCH_DIR=
export REMOTE_EVENT_LOGS=$REMOTE_SCRATCH_DIR/eventlogs/cpu-logs
export REMOTE_CPU_CLUSTER_PROPS=$REMOTE_SCRATCH_DIR/cluster_properties/cpu_cluster_props.json
export REMOTE_OUTPUT_DIR=$REMOTE_SCRATCH_DIR/output_folder

########################################
# running DB-Azure qualification local mode
########################################

# run command with local cpu cluster-properties 
spark_rapids_user_tools databricks_azure qualification \
  --cpu_cluster $CPU_CLUSTER_PROPS \
  --local_folder $LOCAL_OUTPUT_DIR \
  --eventlogs $REMOTE_EVENT_LOGS \
  --remote_folder $REMOTE_OUTPUT_DIR \
  --per-sql \
  --verbose

# run command with remote cpu cluster-properties
spark_rapids_user_tools databricks_azure qualification \
  --cpu_cluster $REMOTE_CPU_CLUSTER_PROPS\
  --local_folder $LOCAL_OUTPUT_DIR \
  --eventlogs $REMOTE_EVENT_LOGS \
  --remote_folder $REMOTE_OUTPUT_DIR \
  --verbose

# run command with remote cpu cluster-properties and no remote output folder
spark_rapids_user_tools databricks_azure qualification \
  --cpu_cluster $REMOTE_CPU_CLUSTER_PROPS\
  --local_folder $LOCAL_OUTPUT_DIR \
  --eventlogs $REMOTE_EVENT_LOGS \
  --verbose

# run command with cpu cluster name
spark_rapids_user_tools databricks_azure qualification \
  --cpu_cluster $CPU_CLUSTER_NAME \
  --local_folder $LOCAL_OUTPUT_DIR \
  --eventlogs $REMOTE_EVENT_LOGS \
  --remote_folder $REMOTE_OUTPUT_DIR \
  --verbose

# Run command with a incorrect eventlogs
spark_rapids_user_tools databricks_azure qualification \
  --cpu_cluster $CPU_CLUSTER_NAME \
  --local_folder $LOCAL_OUTPUT_DIR \
  --eventlogs ${REMOTE_EVENT_LOGS}_ex \
  --remote_folder $REMOTE_OUTPUT_DIR

# run command with GPU cluster name 
# expect no conversion between CPU and GPU
spark_rapids_user_tools databricks_azure qualification \
  --cpu_cluster $CPU_CLUSTER_PROPS \
  --gpu_cluster $GPU_CLUSTER_NAME \
  --local_folder $LOCAL_OUTPUT_DIR \
  --eventlogs $REMOTE_EVENT_LOGS \
  --remote_folder $REMOTE_OUTPUT_DIR \
  --per-sql \
  --verbose

# run command with GPU cluster cluster-properties 
# expect no conversion between CPU and GPU
spark_rapids_user_tools databricks_azure qualification \
  --cpu_cluster $CPU_CLUSTER_PROPS \
  --gpu_cluster $GPU_CLUSTER_PROPS \
  --local_folder $LOCAL_OUTPUT_DIR \
  --eventlogs $REMOTE_EVENT_LOGS \
  --remote_folder $REMOTE_OUTPUT_DIR \
  --per-sql \
  --verbose

################################################
# running DB-Azure qualification local mode with changes in ENV
################################################

# pass non-existing profiles of databricks to test the failure

# non-existing databricks profile
spark_rapids_user_tools databricks_aws qualification \
  --cpu_cluster $CPU_CLUSTER_NAME \
  --local_folder $LOCAL_OUTPUT_DIR \
  --eventlogs $REMOTE_EVENT_LOGS \
  --remote_folder $REMOTE_OUTPUT_DIR \
  --profile non_existing_profile \
  --per-sql \
  --verbose

# pass different profiles of databricks to test the arguments are picked up correctly

# pass invalid credentials to see the error when the tool cannot access the eventlogs


################################################
# running Dataproc qualification local mode
################################################


################################################
# running EMR qualification local mode
################################################


################################################
# running DB-AWS qualification local mode
################################################
```

nartal1 commented:
I haven't gone through it completely yet. Just wanted to know if you are planning to add documentation in this PR or as a follow-on?

cindyyuanjiang commented:
> I haven't gone through it completely yet. Just wanted to know if you are planning to add documentation in this PR or as a follow-on?

I am planning to add documentation in a separate PR. I am putting together some documentation to run the tool and could help with reviewing this PR. I will share with you once that is done. Thanks!

amahussein reviewed (changes requested):
I defined a new profile for databricks-azure
```
spark_rapids_user_tools databricks_azure qualification   \
--cpu_cluster $CPU_CLUSTER_NAME   \
--local_folder $LOCAL_OUTPUT_DIR  \
--eventlogs $REMOTE_EVENT_LOGS  \
--remote_folder $REMOTE_OUTPUT_DIR   \
--profile $PROFILE   \
--verbose
```


The access-key is leaked in the error. We need to prevent the key from showing up in the error message. I am not sure why it is failing though. the access-key seems to be correct.

```
RuntimeError: Error invoking CMD <java -Xmx24g -Drapids.tools.hadoop.fs.azure.account.key.databricksazuretest.dfs.core.windows.net=*****>:
        | Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
        | Exception in thread "main" Failure to initialize configuration
        |       at org.apache.hadoop.fs.azurebfs.services.SimpleKeyProvider.getStorageAccountKey(SimpleKeyProvider.java:51)
        |       at org.apache.hadoop.fs.azurebfs.AbfsConfiguration.getStorageAccountKey(AbfsConfiguration.java:586)
        |       at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystemStore.initializeClient(AzureBlobFileSystemStore.java:1560)
        |       at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystemStore.<init>(AzureBlobFileSystemStore.java:236)
        |       at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem.initialize(AzureBlobFileSystem.java:183)
        |       at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:3469)
        |       at org.apache.hadoop.fs.FileSystem.access$300(FileSystem.java:174)
        |       at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:3574)
        |       at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:3521)
        |       at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:540)
        |       at org.apache.hadoop.fs.Path.getFileSystem(Path.java:365)
        |       at com.nvidia.spark.rapids.tool.EventLogPathProcessor$.getEventLogInfo(EventLogPathProcessor.scala:89)
        |       at com.nvidia.spark.rapids.tool.EventLogPathProcessor$.$anonfun$processAllPaths$1(EventLogPathProcessor.scala:166)
        |       at scala.collection.immutable.List.flatMap(List.scala:366)
        |       at com.nvidia.spark.rapids.tool.EventLogPathProcessor$.processAllPaths(EventLogPathProcessor.scala:166)
        |       at com.nvidia.spark.rapids.tool.qualification.QualificationMain$.mainInternal(QualificationMain.scala:75)
        |       at com.nvidia.spark.rapids.tool.qualification.QualificationMain$.main(QualificationMain.scala:34)
        |       at com.nvidia.spark.rapids.tool.qualification.QualificationMain.main(QualificationMain.scala)

```

amahussein commented:
> I am not sure why it is failing though. the access-key seems to be correct.
> 

The command lines in the PR description are missing `tools_jar` argument since the latest jar on mvn doies not have the Azure connectivity feature yet.

cindyyuanjiang commented:
> The access-key is leaked in the error. We need to prevent the key from showing up in the error message. I am not sure why it is failing though. the access-key seems to be correct.

Updated implementation to prevent key from showing up in error message. Thanks!

amahussein reviewed (changes requested):
@cindyyuanjiang 
It looks like the wrapper is not loading the account-key correctly. It loads the hardcoded storage-account everytime.

1. Make sure that you remove the hardcoded storage-account in `DBAzureLocalRapidsJob._build_jvm_args`
2. Debug why the key is not built out from the eventlogs. After fixing it out test your fix on two different account-storages with different eventlogs patterns. For example, test with a list of eventlogs, and with a path to an abfs directory, single-files..etc.  This will validate that no other hardcoded assumption are left behind.

amahussein commented on user_tools/src/spark_rapids_pytools/cloud_api/databricks_azure_job.py:
```diff
@@ -34,16 +34,21 @@ def get_account_name(cls, url: str):
-                key = conn_str.split('AccountKey=')[1].split(';')[0]
-        if key != '':
-            vm_args.append(f'-Drapids.tools.hadoop.fs.azure.account.key.databricksazuretest.dfs.core.windows.net={key}')
+        elif eventlogs:
```
styling issue: `key` is very common name to be used to retrieve a key-value pair. It will be nice to rename that to something that reflects what the value it represents.

In this context, it looks like `self.exec_ctxt.platform.ctxt['key']`  has precedence on extracting account-name from eventlogs.
What sets `self.exec_ctxt.platform.ctxt['key']`? is that an input from the user?

cindyyuanjiang replied:
Yes, I agree. In azure cli, we can set the access key using `az config set storage.key=xxxxxx`. In the config file `~/.azure/config`, it would be under section `storage`, and key = `key`. I used 'key' mostly to conform with this azure cli convention.
If the user has set the access key using the azure cli command as above, then in `PlatformBase.load_from_config_parser` called by `PlatformBase._set_credential_properties` , this storage 'key' key-value pair would be picked up from the config file and stored in `self.exec_ctxt.platform.ctxt`.

amahussein replied:
> Yes, I agree. In azure cli, we can set the access key using `az config set storage.key=xxxxxx`. In the config file `~/.azure/config`, it would be under section `storage`, and key = `key`. I used 'key' mostly to conform with this azure cli convention. If the user has set the access key using the azure cli command as above, then in `PlatformBase.load_from_config_parser` called by `PlatformBase._set_credential_properties` , this storage 'key' key-value pair would be picked up from the config file and stored in `self.exec_ctxt.platform.ctxt`.

I wonder if we need to use both methods in the wrapper.

Let's think about the following scenario:
- a user installs Azure SDK and he runs `az config set storage.key=xxxxxx` to set her/his key for a storage-account "STORAGE-A"
- User starts to explore our wrapper. It has been a while that she/he forgets that `az config set storage.key=xxxxxx` was executed.
- Because  she/he been using the Azure SDK for sometime, she/he thinks no need to change anything to the configurations configurations.
- she/he runs the wrapper command; but the eventlog is stored on another "STORAGE-B"
- The wrapper loads the key defined in the Azure config file to authenticate .
- The Scala module fails because it tries to authenticate "STORAGE-B" using "STORAGE-A" key.
- It won't be easy to troubleshoot the problem because we redact the actual key value. 

On the other hand, if we rely on a single source to load the key  which is the connection-string, then we are confident that we are loading the correct key for the eventlog (assuming they are all belong to same account-storage). WDYT?

cindyyuanjiang replied:
I think relying on a single source would make it easier for the user, given the scenario you pointed out above. So the user don't need to worry about setting the configuration or looking for the key in the Azure portal. I will update the implementation and test this.

cindyyuanjiang commented:
We addressed the following issues in the latest update:

- addressed issue with duplicated output directories for `--remote-folder`
- appended storage account access keys to `az storage` CLI commands, and redact access key values in logging
- updated `DBAzureLocalRapidsJob.get_account_name` to deal with cases for empty eventlogs list and local eventlogs

amahussein commented:
Followup issue to create documentation for Azure-Databricks usage https://github.com/NVIDIA/spark-rapids-tools/issues/374

amahussein commented on user_tools/src/spark_rapids_pytools/pricing/databricks_azure_pricing.py:
```diff
@@ -72,8 +72,15 @@ def get_ram_size_for_vm(self, machine_type: str) -> str:
+            return rate_per_hour
+        except Exception as ex:  # pylint: disable=broad-except
+            self.logger.info('Encountered exception when retrieving price for instance %s: %s', instance, ex)
+            return -1.0
```
What does the negative rate represent? I am wondering how it looks displaying negative numbers as cost.
As we are comparing costs, the savings will be negated even when they are less expensive on GPU side.
Should we just throw an exception if we can't get the actual rate?

amahussein reviewed (commented):
Is there an example that would not be able to get the cost?

cindyyuanjiang commented:
Follow up issue to include scripts used to process Databricks Azure pricing is tracked here: https://github.com/NVIDIA/spark-rapids-tools/issues/378
